### PR TITLE
Hpc2 theory

### PR DESCRIPTION
**Pull request by NGRhodes**
Please review and provide feedback.
This is still work is progress so not yet ready for merging.
I think all the needed content is included but needs work on wording and formatting.
So far, I have identified the following items to look at:

- [x] Possibly split info on Linux (from HPC) into own section
- [x] HPC0/HPC1 Recap is too wordy
- [ ] Formatting/layout to better match carpentries.
- [x] Environments - does this need to have a better title/name ?
- [x] Where to location section on file types ?
- [x] Does section on build tools need info on Packaging / Testing / Installing
- [ ] Check/add references
- [ ] Update Overview of topics (sections)
- [ ] Summary

### CODE REVIEW

**Comment by jhodrien** [comment]
> * [ ]  HPC0/HPC1 Recap is too wordy

Might be tempted to pretty much scrap the HPC0/HPC1 section entirely?  There's not actually anything in this course that requires you to understand what an HPC is, is there?

> * [ ]  Formatting/layout to better match carpentries.

I'll not pretend to have tried to match carpentries too much, but we need all the sections to be consistent.  We can hammer that out last, as I don't think that's going to take long to sort out.

> * [ ]  Environments - does this need to have a better title/name ?

I'm not sure.  It's probably a Platform more than an Environment?  If you're going to list physical and virtual machines as types of machines, I'd probably make the headings match.

> * [ ]  Where to location section on file types ?

Not sure.  Not sure whether I like the definition of libraries.  It rather mixes a traditional .a/.so type library, from a library of stuff (say what might live in /usr/share).

> * [ ]  Does section on build tools need info on Packaging / Testing / Installing

Perhaps we should have something nodding at rpm/dpkg and pointint out that we're not really playing at the system package level (with a little nod towards containers).

I'll not bother reviewing for niggles at this stage.

**Comment by Sparrow0hawk** [comment]
I agree with @jhodrien I think we can scrap the HPC recap as it probably doesn't add much (attendees of this course should have a rough understanding of HPC and a recap is probably unnecessary for the course content). This section should probably become more a "what is software/ what do we mean by installing software" with the basics of:
- different types of files (source code, libraries, executables etc)
- Where do I put my installed software? (explanation of standard locations in Linux, importance of variables like `PATH` `LD_LIBRARY_PATH`
- Why we can't just install everyone at once
>     * [ ]  HPC0/HPC1 Recap is too wordy

I think current formatting is fine in line with other sections in this book.
>     * [ ]  Formatting/layout to better match carpentries.

Let's call them Platforms as @jhodrien suggests
>     * [ ]  Environments - does this need to have a better title/name ?

Addressed above.
>     * [ ]  Where to location section on file types ?

I think we can make a reference to this, ties in with points above about a "What is software" section.
>     * [ ]  Does section on build tools need info on Packaging / Testing / Installing

Could we add all references/further reading links to the buttom, with subheading to relate to relevant section
>     * [ ]  Check/add references

File: book/course/theory.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+* Package Managers
+* Platforms
+
+### Types of files (WHERE DOES THIS FIT ???????????????)
+
+There are numerous file types you need to be aware of when using and developing code for HPC.
+Below are the key types of files:
+
+* executables - these types of files contain compiled binary code. When opened, the operating system will execute the instructions of the code to run the program.
+* Libraries - these files usually contain compiled code, but cannot be directly run (they can also contain data, templates, documentation). They provide interfaces to be accessed from an existing running program. Libraries are useful for sharing common functionality between different programs.
```
Comment (jhodrien) [inline book/course/theory.md:22]: Not sure I like this definition, although I think the additional note that the term often gets used to cover other things is valid.  I think in terms of building software, I'd go with more of a:

A Library in Linux
A library is a collection of pre-compiled pieces of code called functions. The library contains common functions and together, they form a package called — a library.

Or maybe:

In programming, a library is an assortment of pre-compiled pieces of code that can be reused in a program.

**Review by jhodrien — COMMENTED** [review]
Can we also trim out anything from there that's more an editorial note, as these are liable to get left in otherwise.

File: book/course/theory.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+It is possible to create and run multiple guest machines on a single host.
+
+The guest machines can use different virtual hardware to the host phyiscal machine and any operating system that supports the virtual harware.
+
+### Containers
+
+Containers are an Operating System level virtualisation that provides isolated instances.
+They utilise the host operating sytem's kernel and access to (usually a limited set of) hardware resources (memory, CPU, storage, input devices).
+
+Containers have less overheard (often no noticable performance loss) due to not needing to emulate any hardware, more flexible and require less storage and configuration.
```
Comment (jhodrien) [inline book/course/theory.md:173]:
```suggestion
Containers have less overhead (often no noticeable performance loss) due to not needing to emulate any hardware, are more flexible and require less storage and configuration.
```

File: book/course/theory.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+ARC3 and ARC4 support Singularity and Apptainer (the new sucessor to Singularity) container tools. They support the import and running of Docker containers.
+
+### Virtual Environments
+
+Virtual Environments are environments for managing isolated sets of user software (tools and libraries).
+These are usually managed by a tool such as Anaconda or virtualenv.
+
+Each environemt created can have its own set of software and when combined with package management can specify specific (or minimum or maximum) versions of some (or all) applications and libraries.
+
+ARC3 and ARC4 provides Anaconda as the primary tool for managing environments, others (such as virtualenv for python) as available or installable.
```
Comment (jhodrien) [inline book/course/theory.md:184]:
```suggestion
ARC3 and ARC4 provides Anaconda as the primary tool for managing environments, although others (such as virtualenv for Python) are available or installable.
```

**Review by jhodrien — COMMENTED** [review]
General typos and tidies.

File: book/course/theory.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+An interpreter reads and executes source code directly. They generally runs code slower than a compiled executable as it has to compile code every time the code is run, but there are optimisations such as bytecode compilation and caching that can vastly improve performance.
+
+Python is a popular interpreted language and the most common interpreter is the default, [CPython](https://en.wikipedia.org/wiki/CPython).
+
+Some languages are capable of being compiled or run through an interpreter.
+
+### Build Configuration
+
+Tools like CMake can be used to create different platform specific for different build tools configurations.
```
Comment (Sparrow0hawk) [inline book/course/theory.md:128]:
```suggestion
Tools like CMake can be used to create different platform specific for different build tools configurations.
```
This doesn't make sense, think you're missing a word after "different platform specific"

File: book/course/theory.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+Configuration files usually contain locations of the source code, output locations, of existing components (libraries, assets such as help documentation, data).
+
+They can contain options for conditional builds such as for debug and release builds and for additional features to take advantage of optional components such as input/output format/interfaces or hardware - such as GPUs.
+
+### Packaging
+
+```{Note}
+Users do not have sufficent permissions to install system packages on our HPC systems.
```
Comment (Sparrow0hawk) [inline book/course/theory.md:137]:
```suggestion
Users do not have sufficient permissions to install system packages on our HPC systems.
```

File: book/course/theory.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+Containers provide a mechanism for users to install system packages in a user managed environment.
+```
+
+Linux and package distribtions usually provide tools to automate many steps of the process of creating software packages from source; this will include steps to compile source code, package the files required and test the package is installable/removable.
```
Comment (Sparrow0hawk) [inline book/course/theory.md:141]:
```suggestion
Linux and package distributions usually provide tools to automate many steps of the process of creating software packages from source; this will include steps to compile source code, package the files required and test the package is installable/removable.
```

File: book/course/theory.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+### Distribution
+
+There is usually little need to distribute built academic code.
+
+If you want to distribute your source code, you can use online repositories such as GitHub. These can be linked into open research repositories like Zenodo.
```
Comment (Sparrow0hawk) [inline book/course/theory.md:147]:
```suggestion
If you want to distribute your source code, you can use online repositories such as [GitHub](https://github.com/). These can be linked into open research repositories like [Zenodo](https://zenodo.org/).
```

File: book/course/theory.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+Virtual machines refer to applications (and system level hypervisors) that provide a software representation of hardware to run guest machines.
+
+It is possible to create and run multiple guest machines on a single host.
+
+The guest machines can use different virtual hardware to the host phyiscal machine and any operating system that supports the virtual hardware.
+
+### Containers
+
+Containers are an Operating System level virtualisation that provides isolated instances.
```
Comment (Sparrow0hawk) [inline book/course/theory.md:181]:
```suggestion
Containers are an operating system level virtualisation that provides isolated instances.
```

File: book/course/theory.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+Containers have less overhead (often no noticeable performance loss) due to not needing to emulate any hardware, are more flexible and require less storage and configuration.
+
+ARC3 and ARC4 support Singularity and Apptainer (the new successor to Singularity) container tools. They support the import and running of Docker containers.
+
+### Virtual Environments
+
+Virtual Environments are environments for managing isolated sets of user software (tools and libraries).
+These are usually managed by a tool such as Anaconda or virtualenv.
+
+Each environemt created can have its own set of software and when combined with package management can specify specific (or minimum or maximum) versions of some (or all) applications and libraries.
```
Comment (Sparrow0hawk) [inline book/course/theory.md:193]:
```suggestion
Each environment created can have its own set of software and when combined with package management can specify specific (or minimum or maximum) versions of some (or all) applications and libraries.
```

File: book/course/theory.md [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+ARC3 and ARC4 provides Anaconda as the primary tool for managing environments, although others (such as virtualenv for Python) are available or installable.
```
Comment (Sparrow0hawk) [inline book/course/theory.md:195]:
```suggestion
ARC3 and ARC4 provides the Anaconda module (that contains `conda` as the primary tool for managing environments, although others (such as virtualenv for Python) are available or installable.
```

**Review by Sparrow0hawk — CHANGES_REQUESTED** [review]
Some minor corrections but once resolve happy to approve.

File: book/course/theory.md [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+Package managers are used to manage applications, tools and libraries as packages.
+
+Packages are archives (often compressed files like zip) and can contain any combination of software, libraries and their metadata.
+
+Packages are collected and maintained together as distributions (eg conda-forge).
+
+Packages managers keep track of the packages (and their versions/varients) available, installed and their dependencies.
+
+  they usually allow the following functionality:
+
```
[the remark this answers is not in the document]
Comment (NGRhodes) [reply book/course/theory.md:90]: I agree with @jhodrien on this, in bullet points full sentences should be punctuated.
